ext/datadog_profiling_native_extension/heap_recorder.c

```diff
@@ -525,8 +539,17 @@ static int st_object_records_iterate(DDTRACE_UNUSED st_data_t key, st_data_t val
   const heap_stack *stack = record->heap_record->stack;
   iteration_context *context = (iteration_context*) extra;
 
-  ddog_prof_Location *locations = context->heap_recorder->reusable_locations;
+  const heap_recorder *recorder = context->heap_recorder;
+  size_t iteration_gen = recorder->iteration_gen;
+  size_t min_gen_age = recorder->iteration_min_gen_age;
+  size_t obj_gen_age = iteration_gen - record->object_data.alloc_gen;
+
+  if (obj_gen_age < min_gen_age) {
```

Comment:
So I'm not sure if this could happen _right now_ but it does make me slightly fearful to see this subtraction

```
  size_t obj_gen_age = iteration_gen - record->object_data.alloc_gen;
```

given that `size_t` is supposed to be unsigned.

That is -- what if we get really unlucky and an extra object gets allocated and sampled after we set `iteration_gen`: that would mean an overflow in `obj_gen_age`. I guess that would mean the object would be included, which is not that bad, but... it seems like a nice footgun lol.

Reply:
Hmm. The preparation happens under the GVL and other than the object record snapshot it's not explicitly allocating anything. But yes you're right, even just the new st_table will likely trigger an allocation sample and would be sufficient to cause the situation you describe. Will adjust 👍